File: examples/fetch-prices.js
/**
 * OpenVan.camp Fuel Prices API — Node.js examples
 * Docs: https://github.com/openvan-camp/openvan-camp-public-api
 * API:  https://openvan.camp/api/fuel/prices
 */

const API_BASE = 'https://openvan.camp';

// -------------------------------------------------------------------
// Example 1: Fetch all prices and print a summary
// -------------------------------------------------------------------
async function printPriceSummary() {
  const res = await fetch(`${API_BASE}/api/fuel/prices`);
  const { data, meta } = await res.json();

  console.log(`\n=== Fuel Prices — ${meta.total_countries} countries ===`);
  console.log(`Updated: ${meta.updated_at}\n`);

  // Print Europe diesel prices, sorted cheapest first
  const europeDiesel = Object.values(data)
    .filter(c => c.region === 'europe' && c.prices.diesel !== null)
    .sort((a, b) => a.prices.diesel - b.prices.diesel);

  console.log('Cheapest diesel in Europe (local currency/liter):');
  europeDiesel.slice(0, 10).forEach(c => {
    const change = c.price_changes.diesel;
    const arrow = change > 0 ? '↑' : change < 0 ? '↓' : '→';
    console.log(`  ${c.country_name.padEnd(20)} ${c.prices.diesel.toFixed(3)} ${c.currency} ${arrow}`);
  });
}

// -------------------------------------------------------------------
// Example 2: Fetch prices + currency rates, convert to USD
// -------------------------------------------------------------------
async function pricesInUSD(countryCodes) {
  const [priceRes, rateRes] = await Promise.all([
    fetch(`${API_BASE}/api/fuel/prices`),
    fetch(`${API_BASE}/api/currency/rates`),
  ]);
  const { data } = await priceRes.json();
  const { rates } = await rateRes.json();

  console.log('\n=== Gasoline prices in USD/liter ===');

  for (const code of countryCodes) {
    const c = data[code];
    if (!c || c.prices.gasoline === null) continue;

    // Convert to EUR, then to USD
    const eurPrice = c.prices.gasoline / rates[c.currency];
    let usdPrice = eurPrice * rates['USD'];

    // If country uses gallons, convert to liters
    if (c.unit === 'gallon') usdPrice /= 3.78541;

    console.log(`  ${c.country_name.padEnd(20)} $${usdPrice.toFixed(3)}/L`);
  }
}

// -------------------------------------------------------------------
// Example 3: Find countries where LPG is available and cheap
// -------------------------------------------------------------------
async function cheapLPG() {
  const [priceRes, rateRes] = await Promise.all([
    fetch(`${API_BASE}/api/fuel/prices`),
    fetch(`${API_BASE}/api/currency/rates`),
  ]);
  const { data } = await priceRes.json();
  const { rates } = await rateRes.json();

  const countries = Object.values(data)
    .filter(c => c.prices.lpg !== null && !c.is_excluded)
    .map(c => ({
      name: c.country_name,
      lpgEur: c.prices.lpg / rates[c.currency],
      currency: c.currency,
      unit: c.unit,
    }))
    .sort((a, b) => a.lpgEur - b.lpgEur);

  console.log('\n=== Cheapest LPG (EUR/liter equivalent) ===');
  countries.slice(0, 10).forEach(c => {
    console.log(`  ${c.name.padEnd(25)} €${c.lpgEur.toFixed(3)}/L`);
  });
}

// -------------------------------------------------------------------
// Run all examples
// -------------------------------------------------------------------
(async () => {
  await printPriceSummary();
  await pricesInUSD(['US', 'DE', 'GB', 'TR', 'AU', 'JP']);
  await cheapLPG();
})().catch(console.error);

File: examples/fetch-prices.py
"""
OpenVan.camp Fuel Prices API — Python examples
Docs: https://github.com/openvan-camp/openvan-camp-public-api
API:  https://openvan.camp/api/fuel/prices

Requirements: pip install requests
"""

import requests

API_BASE = "https://openvan.camp"


def get_prices():
    """Fetch current fuel prices for all countries."""
    r = requests.get(f"{API_BASE}/api/fuel/prices", timeout=10)
    r.raise_for_status()
    return r.json()


def get_rates():
    """Fetch current currency exchange rates (EUR base)."""
    r = requests.get(f"{API_BASE}/api/currency/rates", timeout=10)
    r.raise_for_status()
    return r.json()["rates"]


# -------------------------------------------------------------------
# Example 1: Print all European diesel prices
# -------------------------------------------------------------------
def example_europe_diesel():
    response = get_prices()
    data = response["data"]

    europe = [
        (info["country_name"], info["prices"]["diesel"], info["currency"])
        for code, info in data.items()
        if info["region"] == "europe" and info["prices"]["diesel"] is not None
    ]
    europe.sort(key=lambda x: x[1])

    print(f"\n=== Diesel prices in Europe ({response['meta']['total_countries']} countries total) ===")
    for name, price, currency in europe:
        print(f"  {name:<25} {price:.3f} {currency}/liter")


# -------------------------------------------------------------------
# Example 2: Convert to a target currency using rates API
# -------------------------------------------------------------------
def convert_to_currency(price, from_currency, to_currency, rates, from_unit="liter"):
    """Convert price from local currency to target currency, normalized to liter."""
    if price is None:
        return None
    price_eur = price / rates[from_currency]
    if from_unit == "gallon":
        price_eur /= 3.78541
    return price_eur * rates[to_currency]


def example_prices_in_usd():
    data = get_prices()["data"]
    rates = get_rates()

    country_codes = ["US", "DE", "GB", "TR", "AU", "BR", "JP", "ZA"]

    print("\n=== Gasoline prices in USD/liter ===")
    for code in country_codes:
        c = data.get(code)
        if not c or c["prices"]["gasoline"] is None:
            continue
        usd = convert_to_currency(
            c["prices"]["gasoline"],
            c["currency"],
            "USD",
            rates,
            from_unit=c["unit"],
        )
        print(f"  {c['country_name']:<25} ${usd:.3f}/L")


# -------------------------------------------------------------------
# Example 3: Build a Pandas DataFrame
# -------------------------------------------------------------------
def example_dataframe():
    try:
        import pandas as pd
    except ImportError:
        print("\n(pandas not installed — skipping dataframe example)")
        return

    data = get_prices()["data"]
    rates = get_rates()

    rows = []
    for code, c in data.items():
        for fuel in ("gasoline", "diesel", "lpg"):
            price = c["prices"].get(fuel)
            if price is None:
                continue
            eur = price / rates[c["currency"]]
            if c["unit"] == "gallon":
                eur /= 3.78541
            rows.append({
                "country_code": code,
                "country": c["country_name"],
                "region": c["region"],
                "fuel": fuel,
                "price_local": price,
                "currency": c["currency"],
                "price_eur_per_liter": round(eur, 4),
                "updated": c["fetched_at"],
            })

    df = pd.DataFrame(rows)
    print("\n=== DataFrame (first 10 rows, diesel in Europe) ===")
    print(
        df[(df["region"] == "europe") & (df["fuel"] == "diesel")]
        .sort_values("price_eur_per_liter")
        .head(10)
        .to_string(index=False)
    )
    return df


# -------------------------------------------------------------------
# Run all examples
# -------------------------------------------------------------------
if __name__ == "__main__":
    example_europe_diesel()
    example_prices_in_usd()
    example_dataframe()
